## Setup and Imports

In [1]:
# Automatically reload our modules if we change them in VSCode/PyCharm
%load_ext autoreload
%autoreload 2

# Add the project root to the path so we can import from src/
import sys
from pathlib import Path
sys.path.append(str(Path.cwd().parent.parent))

from src.data.loader import load_raw_data
from src.utils.profiler import generate_data_profile

## Load the Data

In [2]:
# Load dataset
df_titanic = load_raw_data("Titanic.csv")

print(f"Dataset Shape: {df_titanic.shape}")
print(f"Total Duplicate Rows: {df_titanic.duplicated().sum()}")
df_titanic.head(3)

Dataset Shape: (891, 12)
Total Duplicate Rows: 0


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


## Generate Data Profile

In [3]:
# Run our custom profiler
profile = generate_data_profile(df_titanic)
profile

,Data Type,Non-Null Count,Missing Values,Missing Percentage (%),Unique Values
PassengerId,int64,891,0,0.00,891
Survived,int64,891,0,0.00,2
Pclass,int64,891,0,0.00,3
Name,object,891,0,0.00,891
Sex,object,891,0,0.00,2
Age,float64,714,177,19.87,88
SibSp,int64,891,0,0.00,7
Parch,int64,891,0,0.00,7
Ticket,object,891,0,0.00,681
Fare,float64,891,0,0.00,248


## Phase 2 Conclusion: Cleaning Plan

**Findings:**
- The dataset contains 891 rows and 12 columns.
- Missing values identified in: `Cabin` (77.10%), `Age` (19.87%), and `Embarked` (0.22%).
- No duplicate rows detected.

**Action Plan for Phase 3 (Cleaning):**
1. **Cabin:** Drop the entire column. 77% missing data is too high a threshold for reliable imputation.
2. **Embarked:** Drop the 2 rows with missing values. The loss of data is negligible (< 0.3%).
3. **Age:** **DO NOT IMPUTE.** 
   - *Architectural Note on Data Leakage:* Imputing missing values with the mean/median of the entire dataset right now will cause data leakage into our future test set. 
   - Pandas/Seaborn can natively handle `NaN` values during Exploratory Data Analysis (EDA). We will leave the `NaN`s intact for EDA and handle the imputation strictly within the Scikit-Learn preprocessing pipeline in Week 4.